_posts/2022-08-22-fan-out-and-back-in-using-functions.md

```diff
@@ -0,0 +1,415 @@
+
+```python
+redisClient = None
+
+def initRedis():
+    redisHostname = os.getenv('redis_hostname')
+    redisPort = os.getenv('redis_port')
+
+    with open('/var/openfaas/secrets/redis-password', 'r') as s:
+        redisPassword = s.read()
+
+    return redis.Redis(
+        host=redisHostname,
+        port=redisPort,
+        password=redisPassword,
+    )
+```
+
+Initialize the redis client in the function handler:
+
+```python
+def handle(event, context):
+    global redisClient
+
+    if redisClient == None:
+        redisClient = initRedis()
+```
+
+Update the stack.yaml for the functions to add the Redis environment variables and secret: 
+```yaml
+environment:
+    s3_bucket: of-demo-inception-data
+    redis_hostname: "redis-master.redis.svc.cluster.local"
+    redis_port: 6379
+secrets:
+    - s3-key
+    - s3-secret
+    - redis-password
+```
+
+Don't forget to add `redis` to the `requirements.txt` for your functions.
+
+The `create-batch` function can now be updated to count the number of URLs in the CSV file and initialize a Redis key for `batchId` with that value. 
+```python
+batchFile = event.body.decode()
+s3URL = 's3://{}/{}'.format(bucketName, batchFile)
+with open(s3URL, 'rb', transport_params={'client': session.client('s3')}) as f:
+    records = pd.read_csv(f)
+
+batchId = str(uuid.uuid4())
+batchSize = len(records)
+
+# Initialize a counter in Redis
+redisClient.set(batchId, batchSize)
+
+for index, col in records.iterrows():
+    headers = { 'X-Batch-Id': batchId }
+    res = requests.post('http://gateway.openfaas:8080/async-function/run-model', data=col['url'], headers=headers)
+```
+
+In the `run-model` function we need to decrement this counter every time an invocation to the `inception` function finishes. If the counter reaches zero, this means all work in the batch is completed and the final function can be invoked.
+
+```python
+
+remainingWork = redisClient.decr(batchId)
+
+if remainingWork == 0:
+    headers = { 'X-Batch-Id': batchId }
+    res = requests.post("http://gateway.openfaas:8080/async-function/collect-result", headers=headers)
+    r.delete(batchId)
+```
+
+### Collect the results
+The final function is called when all asynchronous requests in a batch are completed. This function can be used to aggregate the results, notify some other actor or trigger the next step in your processing workflow. In this example we are going to use it to collect all the responses in a single json file and upload that file to the S3 bucket.
+
+The code of the `collect-result` function:
+
+```python
+import os
+import json
+import boto3
+
+from smart_open import s3
+from smart_open import open
+
+def handle(event, context):
+    bucketName = os.getenv('s3_bucket')
+
+    with open('/var/openfaas/secrets/s3-key', 'r') as s:
+        s3Key = s.read()
+    with open('/var/openfaas/secrets/s3-secret', 'r') as s:
+        s3Secret = s.read()
+
+    session = boto3.Session(
```

Comment:
Am I remembering correctly, didn't you optimise this to run once somewhere and in doing so reduced the duration taken for each request?

Reply:
This was done in the other functions but not for this one yet.

I use `s3.iter_bucket` from the `smart_open` module to iterate over the S3 bucket’s contents and I don't think it supports passing in the client. That is why the function was not updated yet.

Reply:
The secrets probably don't have to be read for every invocation though?

Reply:
The code has been updated.